pennylane/wires.py

```diff
@@ -0,0 +1,399 @@
+from numbers import Number
+import numpy as np  # for random functions
+
+
+class WireError(Exception):
+    """Exception raised by a :class:`~.pennylane.wires.Wire` object when it is unable to process wires.
+    """
+
+
+class Wires(Sequence):
+    """
+    A bookkeeping class for wires, which are ordered collections of unique objects. The i'th object represents
+    the i'th quantum subsystem.
+
+    There is no conceptual difference between registers of multiple wires and single wires,
+    which are just wire registers of length one.
+
+    Indexing and slicing this sequence will return another ``Wires`` object.
+
+    Args:
+         wires (Any): If iterable, interpreted as an ordered collection of unique objects representing wires.
+            Any other input type is converted into an iterable of a single entry, and hence interpreted as a
+            single wire.
+    """
+
+    def __init__(self, wires):
+
+        if isinstance(wires, Wires):
+            # If input is already a Wires object, just copy its wire list
+            self.wire_list = wires.wire_list
+
+        elif isinstance(wires, Iterable):
+            # If input is an iterable, check that entries are unique and convert to a list
+            try:
+                if len(set(wires)) != len(wires):
+                    raise WireError(
+                        "Wires must be unique; got {}.".format(wires)
+                    )
+            except TypeError:
+                raise WireError("Cannot create wires from elements of type {}.".format(type(wires[0]))) # TODO: edge case that iterable contains different types
+
+            self.wire_list = list(wires)
+
+        else:
+            # If input is not an iterable, interpret it as an object representing a single wire
+            self.wire_list = [wires]
+
+    def __getitem__(self, idx):
+        """Method to support indexing. Returns a Wires object representing a register with a single wire."""
+        return Wires(self.wire_list[idx])
+
+    def __len__(self):
+        """Method to support ``len()``."""
+        return len(self.wire_list)
+
+    def __contains__(self, item):
+        """Method checking if Wires object contains an object."""
+        if isinstance(item, Wires):
+            # If all wires can be found in this object, return True
+            if all(wire in self.wire_list for wire in item.wire_list):
+                return True
+        return False
+
+    def __repr__(self):
+        """Method defining the string representation of this class."""
+        return "<Wires = {}>".format(self.wire_list)
+
+    def __eq__(self, other):
+        """Method to support the '==' operator. This will also implicitly define the '!=' operator."""
+        if isinstance(other, self.__class__):
+            return self.wire_list == other.wire_list
+        return False
+
+    def __hash__(self):
+        """Implements the hash function, used for example by ``set()``."""
+        # hash the string representation
```

Comment:
Note that if the object had an internal `frozenset` representation of its wires (in addition to its `wires_list`), then hashing could be done directly on that

Reply:
But that would mean Wires([1, 2, 3]) would hash to the same value as Wires([2, 1, 3])? Well, makes sense!